### DISCUSSION TITLE
Web API 2.0

### DISCUSSION THREAD

**Discussion by WedgeSparda** [RetroAchievements/RAWeb#1212] [Ideas]
Hi, recently I started developing a Swift Web API client for RA, and noticed that the current web API lacks some data and is not consistent in data formats.

So I’m thinking of help developing a 2.0 version but don’t know if that’s something already set in the roadmap.

Is there any development regarding this possible feature?

**Comment by luchaos** [comment]
Hey @WedgeSparda - thanks for the input!
The current Web API is lacking indeed :) 
There are plans for creating a v2 for it in the future but have to move some things into place first.
Preparations have been made to support both the current version and the upcoming version on a dedicated subdomain.

**Reply by WedgeSparda** [reply]
Hi @luchaos, great to hear that. If you need some help developing it, let me know. I’m an iOS dev but I have some background in web development.

**Reply by WelingtonMonteiro** [reply]
About a WEB 2.0 API a year ago I created a skeleton in node.js of all the existing modules, if you want, you could start with it, it already includes a crud for all the modules, you would have to include the specific routes of each module.
In addition to already having unit tests for each module:
The project is well organized and divided into modules.

Sequelize configuration, docker, dotenv, routes.
![image](https://user-images.githubusercontent.com/4640319/207619421-78b32fd1-023a-40a3-bf7b-a814ad8fe0d7.png)

File generic crud
![image](https://user-images.githubusercontent.com/4640319/207619640-e7ff08dd-7bfa-483c-ac5a-c9dddcb7fb62.png)

Basic crud generic route, being able to overwrite or add more routes:
![image](https://user-images.githubusercontent.com/4640319/207619966-62dd0db7-b2c3-4b1d-ba63-1fd62a8fa2c9.png)

Basic crud generic controller, being able to overwrite or add more functions
![image](https://user-images.githubusercontent.com/4640319/207620110-0456eb3a-67c5-4de8-bf3d-6a5a2a5d17b8.png)

Basic crud generic service, being able to overwrite or add more functions
![image](https://user-images.githubusercontent.com/4640319/207620162-ee744d53-3e7a-471f-a334-40b806a6dcbd.png)

Basic crud repository controller, being able to overwrite or add more functions
![image](https://user-images.githubusercontent.com/4640319/207620217-4f915942-9316-4e45-8ca1-ac17fb6c1849.png)

Schemas of all site modules exported in sequelize
![image](https://user-images.githubusercontent.com/4640319/207620297-680ddb1a-9dfb-44cf-a8af-c4a24ec1f298.png)

Unit Tests of each modules using mocha:
![image](https://user-images.githubusercontent.com/4640319/207620494-d6d3c16e-11e2-46a5-bb33-0301e169ccaa.png)

At the time, I did it because I wanted to help, and because there was no web api. But at the time they were migrating the site to a 2.0 version of PHP, this version today.

I don't know who is responsible, but if you need it, there's a skeleton to start a web api.

**Reply by WedgeSparda** [reply]
Hi @WelingtonMonteiro thanks for the info.

Is this WEB API 2.0 project in another repo? Can't find it.

**Reply by WelingtonMonteiro** [reply]
I haven't uploaded it to the repo yet, it's in my private repository, I'm waiting for the responsible devs to release it, as they still don't know if they'll use it or create another one from scratch.

**Comment by WedgeSparda** [comment]
@luchaos Have a question about the upcoming version of the API? Is it going to be PHP code too or is it going to be a different technology?

**Reply by WelingtonMonteiro** [reply]
As I mentioned above as a friend a while ago I made a skeleton of an api in nodejs with sequelize, with all the modules available at the time, already implemented basic CRUD, and configurations. If you don't want to start from scratch, here's my contribution.

**Reply by luchaos** [reply]
much appreciated @WelingtonMonteiro - this might come in handy in the future. 
with careful consideration to not to discredit your efforts; not much has changed since our last conversation. 

the goal for now is still to consolidate all features in a monolithic laravel application to better allow us splitting the services first (read: we're not quite there yet). 

an initial reiteration of the web api will most likely still be implemented with php/laravel because this keeps it closer to the current codebase, whilst still allowing us to improve the response formats and URL schemas. after that's done i'm personally very open to changing the stacks of the individual services if they can be integrated easily and are reasonably maintainable - until then i'd rather keep the current setup for our current developers' convenience.

**Reply by WedgeSparda** [reply]
Sorry for my late reply @WelingtonMonteiro 

My question to @luchaos was indeed related to the use of different technologies for 1.0 and 2.0. Since I don't know what plans are for the Web API 2.0 just wanted to know if your Node.js code would be integrated later into the new web. As I said, I stopped working on web development a few years ago and I'm not sure which is the "best" approach.
